### PR TITLE
perf(core): reduce allocations and improve hot-path performance

### PR DESCRIPTION
**Pull request by ckeller42**
## Summary

Algorithmic complexity fix and test infrastructure for the compute-view pipeline.

### What changed
- **Fix O(n²) in `topologicalSort`**: `ensureParentsFirst` used `shift()`+`findIndex()`+`splice()` which is quadratic. Replaced with Map-based O(n) lookups. Same for the post-topsort unsorted-node merge (`indexOf`/`includes` scans replaced with Map). Not noticeable on current workspaces (462 elements), but prevents a scaling wall as models grow.
- **`forEachAncestorFqn` visitor**: zero-allocation ancestor traversal with early-exit, replacing the `[self, ...ancestorsFqn()]` array pattern across 6 call sites
- **Decorate-sort-undecorate** in `buildComputedNodes` to avoid repeated `split(".")` per comparison
- **`findRelations` memoization** in `DynamicViewCompute` (per-compute-run cache)
- **`applyViewRuleStyles`**: plain loop replacing `pipe(filter, filter, forEach)`
- **`WorkspaceManager`**: Map-based file dedup replacing `uniqueBy` array materialization
- **Deprecate** `compareByFqnHierarchically` (zero importers after buildComputedNodes change)
- **Delete** `common-head.spec.ts` (covered by new link-nodes-with-edges.spec.ts)
- Benchmark script for reproducible measurements

### What was tried and reverted
An LCA rewrite in `linkNodesWithEdges` (replacing `ancestorsOfNode`+`commonHead` with a `parentDepth` walk-to-root algorithm) was reverted after isolation testing showed it caused a ~6% regression on large workspaces (872ms vs 866ms baseline). The triple parent-chain traversal per edge outweighed the allocation savings at 3127 relations.

### Performance
Measured performance-neutral on a large workspace (462 elements, 3127 relations, 137 views): 872ms vs 866ms baseline — within noise. The value is preventive: the O(n²) topsort would become a bottleneck as workspace size grows.

### Testing
- 14 new unit tests for `linkNodesWithEdges` (edge.parent LCA, inEdges/outEdges propagation, group-node edge cases)
- 5 new tests for `forEachAncestorFqn` (traversal, root FQN, includeSelf, early-exit)
- CI perf guardrails: algorithmic complexity ratio tests (detect O(n²) regressions) and heap allocation budget tests (detect array-allocation regressions)
- Benchmark script: `pnpm --filter likec4 run bench:workspace`
- All 938 existing tests pass

### CODE REVIEW

**Comment by ckeller42** [comment]
### Code review

No issues found. Checked for bugs and CLAUDE.md compliance.

🤖 Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Comment by ckeller42** [comment]
### Code review

Found 2 issues:

1. No changeset file — `forEachAncestorFqn` is a new public export from `@likec4/core` and the PR delivers measurable user-visible improvements, but there is no `.changeset/*.md` entry. CLAUDE.md says: _"Project requires changeset files (folder `.changeset/`)"_

https://github.com/likec4/likec4/blob/990f3ad07d5dcfa51615dd3d5d8b5535493db7f9/packages/core/src/utils/index.ts#L1-L5

2. `ancestorsOfNode.ts` is dead code not deleted — all import sites were removed by this PR, and the PR's own `TODO-perf.md` acknowledges "_`ancestorsOfNode.ts` is now dead code (no remaining importers)_", but the file was not deleted.

https://github.com/likec4/likec4/blob/990f3ad07d5dcfa51615dd3d5d8b5535493db7f9/packages/core/src/compute-view/utils/ancestorsOfNode.ts#L1-L10

🤖 Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Comment by ckeller42** [comment]
### Code review

No issues found. Checked for bugs and CLAUDE.md compliance.

🤖 Generated with [Claude Code](https://claude.ai/code)

**Comment by ckeller42** [comment]
@coderabbitai review

File: packages/core/src/compute-view/utils/topological-sort.ts [hunk 1]
```diff
@@ -1,41 +1,55 @@
...
+    let parentId = item.parent
+    while (parentId) {
+      const parentItem = remaining.get(parentId)
+      if (!parentItem) break
+      chain.push(parentItem)
+      parentId = parentItem.parent
+    }
+    // Reverse so root ancestors come first
+    return chain.reverse()
+  }
```
[the remark this answers is not in the document]
Comment (ckeller42) [reply packages/core/src/compute-view/utils/topological-sort.ts]: Verified false positive. The precondition is documented in the JSDoc: parent IDs must exist in the array, guaranteed by `buildComputedNodes` which sets `node.parent` to the nearest ancestor in the map. Independent algorithmic review confirmed correctness for all input permutations.

File: packages/likec4/scripts/benchmark-workspace.ts [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    return summarize(samples, args)
+  }
+
+  const likec4 = await LikeC4.fromWorkspace(args.workspace, {
+    logger: false,
+    printErrors: false,
+    throwIfInvalid: true,
+  })
+  try {
+    const projectId = args.project ?? likec4.projects()[0]
```
[the remark this answers is not in the document]
Comment (ckeller42) [reply packages/likec4/scripts/benchmark-workspace.ts:143]: Fixed in 8b1d8d190 — both call sites now throw an explicit error if no projects are found.

File: packages/likec4/scripts/benchmark-workspace.ts [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+
+async function inspectWorkspace(args: Args) {
+  const likec4 = await LikeC4.fromWorkspace(args.workspace, {
+    logger: false,
+    printErrors: false,
+    throwIfInvalid: true,
+  })
+  try {
+    const projectId = args.project ?? likec4.projects()[0]
+    const model = await likec4.computedModel(projectId)
```
[the remark this answers is not in the document]
Comment (ckeller42) [reply packages/likec4/scripts/benchmark-workspace.ts:183]: Fixed in same commit (8b1d8d190).

File: TODO-perf.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
+# Performance TODO
+
+## Done
+
+- Installed and switched local Node to `22.22.1` and verified `pnpm 10.32.1`.
+- Added a repeatable workspace benchmark driver:
+  - [`packages/likec4/scripts/benchmark-workspace.ts`](/home/ckeller/src/likec4/packages/likec4/scripts/benchmark-workspace.ts)
```
[the remark this answers is not in the document]
Comment (ckeller42) [reply TODO-perf.md]: File deleted in 8b1d8d190.

File: TODO-perf.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+- Captured baseline timings for:
+  - `examples/cloud-system`
+  - `~/dagviews_artifacts/artifacts/arch_src/monaco_aml_lidar`
+- Implemented runtime/memory optimizations in shared hot paths:
+  - added `forEachAncestorFqn` to reduce ancestor-array allocation
+  - replaced several `[self, ...ancestorsFqn(...)]` and `ancestorsFqn(...)` loops with iterator-style walks
+  - reduced allocation/pipeline overhead in `applyViewRuleStyles`
+  - rewrote shared edge-linking parent resolution to avoid ancestor-array construction
+- Added tests for the new ancestor helper in:
+  - [`packages/core/src/utils/fqn.spec.ts`](/home/ckeller/src/likec4/packages/core/src/utils/fqn.spec.ts)
```
[the remark this answers is not in the document]
Comment (ckeller42) [reply TODO-perf.md]: File deleted.

File: TODO-perf.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+## Next Bundle / Static App Work
+
+- Add a bundle-size measurement script for static output:
+  - total output size
+  - gzip/brotli sizes
+  - largest JS chunks
+  - initial route payload
+- Revisit chunk splitting in:
+  - [`packages/likec4/src/vite/config-app.prod.ts`](/home/ckeller/src/likec4/packages/likec4/src/vite/config-app.prod.ts)
```
[the remark this answers is not in the document]
Comment (ckeller42) [reply TODO-perf.md]: File deleted.

File: packages/core/src/compute-view/utils/link-nodes-with-edges.spec.ts [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    id: `${source}->${target}` as EdgeId,
+    source: source as scalar.NodeId,
+    target: target as scalar.NodeId,
+    parent: null,
+    label: null,
+    relations: [],
+    color: 'primary',
+    line: 'solid',
+    head: 'normal',
+  } as ComputedEdge
```
[the remark this answers is not in the document]
Comment (ckeller42) [reply packages/core/src/compute-view/utils/link-nodes-with-edges.spec.ts:57]: Pre-existing pattern — the same `as any as ComputedNodeSource` cast exists in the adjacent `topological-sort.spec.ts` (line 22). Both test helpers construct partial mock objects.

**Comment by ckeller42** [comment]
Closing: performance-neutral on large workspaces after isolation testing revealed the LCA rewrite caused a ~6% regression. The O(n²) topsort fix has value but the PR accumulated too much scope. Will revisit as a smaller, focused change if needed.
